### PR TITLE
[Enhancement] Expr enhance with cache hash

### PR DESCRIPTION
**Pull request by before-Sunrise**
Why I'm doing:
Expr like 100 or takes too long in transform phase, Especially NormalizePredicateRule‘s visitCompoundPredicate, because its time complexity is O(n^2) n is tree node numbers.

What I'm doing:
when visit expr tree bottom up, if child is compound operator, cache child's unique compound tree leaf node in set and child's hash value to eliminate duplicate calculations

benchmark with 2000 predicate: 

without any optimization：
`NormalizePredicateBench.bench_NormalizePredicate_Random_Non_MV                  2000  avgt    5   205.584 ±   27.420  ms/op`

without cache hash value:
`NormalizePredicateBench.bench_NormalizePredicate_Random_Non_MV                  2000  avgt    5   195.232 ±    0.440  ms/op`

cache hash value:
`NormalizePredicateBench.bench_NormalizePredicate_Random_Non_MV                  2000  avgt    5   108.491 ±   6.336  ms/op`

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/HashCachedScalarOperator.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+// limitations under the License.
+
+package com.starrocks.sql.optimizer.operator.scalar;
+
+import com.starrocks.sql.optimizer.base.ColumnRefSet;
+
+import java.util.List;
+import java.util.Objects;
+
+public class HashCachedScalarOperator extends ScalarOperator {
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/HashCachedScalarOperator.java:22]: Could you please make this optimization more generic for all rules, not only `hashCode`, but also `equals`, and maybe some other states ?

Comment (before-Sunrise) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/HashCachedScalarOperator.java:22]: No idea how to do this. Wiseman, do you have any advise?

**Review by Seaven — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/CompoundPredicateOperator.java [hunk 2]
```diff
@@ -25,11 +25,24 @@
...
+    // These two filed are used in NormalizePredicateRule to eliminate common CompoundPredicate
+    // For Expr tree like below, the CompoundTreeLeafNodeNumber is 5, and compoundTreeUniqueLeave's size is 4.
+    //           AND
+    //        /        \
+    //      AND         AND
+    //     /   \       /  \
+    // subT1   a+1  And  subT5
+    //               /   \
+    //             subT3  a+1
+    private int compoundTreeLeafNodeNumber;
```
Comment (satanson) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/CompoundPredicateOperator.java:43]: compoundTreeLeafNodeNumber is the same to compoundTreeUniqueLeaves.size(), if so, remove it and use the latter instead.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/HashCachedScalarOperator.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    public ScalarOperator operator;
+    private int hashValue = 0;
```
Comment (satanson) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/HashCachedScalarOperator.java:24]: do you want to cache hashValue, it seems the implementation is error-prone. modify the code as follows
```java
private transient Integer hashValue = null;

private int hashCodeImpl() {
    return ...;// compute hashCode in actual
}
@Override
public int hashCode() {
   if (hashValue==null) {
        hashValue = hashCodeImpl();
   }
   return hashValue;
}
...

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/CompoundPredicateOperator.java (hunk 2 above)
Comment (before-Sunrise) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/CompoundPredicateOperator.java:43]: Not really. compoundTreeLeafNodeNumber is the tree leaf nodes' size, which contains non-unique node, but compoundTreeUniqueLeaves.size() is size of unique leaf node, which is equal or greater tahn compoundTreeLeafNodeNumber. I need to compare compoundTreeLeafNodeNumber with  compoundTreeUniqueLeaves.size() to find out whether this tree can be optimized or not

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/NormalizePredicateRule.java [hunk 4]
```diff
@@ -130,27 +135,74 @@ public ScalarOperator visitBetweenPredicate(BetweenPredicateOperator predicate,
     @Override
     public ScalarOperator visitCompoundPredicate(CompoundPredicateOperator predicate,
                                                  ScalarOperatorRewriteContext context) {
+        if (predicate.isAnd() || predicate.isOr()) {
+            ScalarOperator newTree = getOptimizedCompoundTree(predicate);
```
Comment (satanson) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/NormalizePredicateRule.java:139]: suggest to make getOptimizedCompoundTree return Optional<ScalarOperator> type, if so, this functions can be rewritten more succinctly.
```java
return getOptimizedCompoundTree(predicate).orElse(predicate);
```

**Review by satanson — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Comment by before-Sunrise** [comment]
@Mergify backport branch-3.3

**Comment by before-Sunrise** [comment]
@Mergify backport branch-3.2

**Pull request merged**